### PR TITLE
Make CWS multi-contest

### PR DESCRIPTION
**Pull request by wil93**
With this patch, ContestWebServer does not require a `-c` flag anymore. When you start CWS, it will expose a list of all available contests on `/`, and you can click on the contest that you're interested in.

The former cookie behavior was to save 2 cookies: `login` and `unread_count`. The new cookie behavior is to save 2*N cookies where N is the number of contests. Each cookie will either be `<contest-name>_login` or `<contest-name>_unread_count`.

This PR depends on #545

---

This change is [Reviewable](https://reviewable.io/reviews/cms-dev/cms/594)

### CODE REVIEW

File: cms/server/contest/handlers/taskusertest.py [hunk 1]
```diff
@@ -125,12 +125,11 @@ class UserTestHandler(BaseHandler):
...
     @tornado.web.authenticated
     @actual_phase_required(0)
-    def post(self, task_name):
+    def post(self, contest_name, task_name):
         participation = self.current_user
 
         if not self.r_params["testing_enabled"]:
-            self.redirect("/")
-            return
+            raise tornado.web.HTTPError(403)
```
Comment (wil93) [inline cms/server/contest/handlers/taskusertest.py]: This change is not needed for the overall patch, but I figured it was a sensible change (which didn't require a different commit or a different PR)

File: cms/server/contest/handlers/taskusertest.py [hunk 2]
```diff
@@ -438,16 +439,16 @@ def post(self, task_name):
...
 
 
-class UserTestStatusHandler(BaseHandler):
+class UserTestStatusHandler(ContestHandler):
 
     refresh_cookie = False
 
     @tornado.web.authenticated
     @actual_phase_required(0)
-    def get(self, task_name, user_test_num):
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/taskusertest.py]: Codacy Issue found: [Arguments number differs from overridden 'get' method](https://www.codacy.com/app/s-maggiolo/cms/file/2347582229/issues/source?bid=3167167&fileBranchId=3216584#l451)

File: cms/server/contest/handlers/taskusertest.py [hunk 3]
```diff
@@ -51,25 +51,24 @@
...
 
-class UserTestInterfaceHandler(BaseHandler):
+class UserTestInterfaceHandler(ContestHandler):
     """Serve the interface to test programs.
 
     """
     @tornado.web.authenticated
     @actual_phase_required(0)
-    def get(self):
+    def get(self, contest_name):
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/taskusertest.py]: Codacy Issue found: [Arguments number differs from overridden 'get' method](https://www.codacy.com/app/s-maggiolo/cms/file/2347582229/issues/source?bid=3167167&fileBranchId=3216584#l67)

File: cms/server/contest/handlers/contest.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        if self.get_secure_cookie(cookie_name) is None:
+            return None
+
+        # Parse cookie.
+        try:
+            cookie = pickle.loads(self.get_secure_cookie(cookie_name))
+            username = cookie[0]
+            password = cookie[1]
+            last_update = make_datetime(cookie[2])
+        except:
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/contest.py]: Codacy Issue found: [No exception type(s) specified](https://www.codacy.com/app/s-maggiolo/cms/file/2347582225/issues/source?bid=3167167&fileBranchId=3216584#l204)

File: cms/server/contest/handlers/contest.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+        self.langs = self.application.service.langs
+        lang_codes = self.langs.keys()
+
+        if self.contest and len(self.contest.allowed_localizations) > 0:
+            lang_codes = filter_language_codes(
+                lang_codes, self.contest.allowed_localizations)
+
+        # Select the one the user likes most.
+        basic_lang = lang_codes[0].replace("_", "-") \
+            if self.contest and len(self.contest.allowed_localizations) else 'en'
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/contest.py]: Codacy Issue found: [Line too long (81/80)](https://www.codacy.com/app/s-maggiolo/cms/file/2347582225/issues/source?bid=3167167&fileBranchId=3216584#l277)

File: cms/server/contest/handlers/tasksubmission.py [hunk 6]
```diff
@@ -359,18 +361,16 @@ def post(self, task_name):
...
 
 
-class TaskSubmissionsHandler(BaseHandler):
+class TaskSubmissionsHandler(ContestHandler):
     """Shows the data of a task in the contest.
 
     """
     @tornado.web.authenticated
     @actual_phase_required(0)
-    def get(self, task_name):
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/tasksubmission.py]: Codacy Issue found: [Arguments number differs from overridden 'get' method](https://www.codacy.com/app/s-maggiolo/cms/file/2347582228/issues/source?bid=3167167&fileBranchId=3216584#l373)

File: cms/server/contest/handlers/tasksubmission.py [hunk 7]
```diff
@@ -418,13 +418,13 @@ def get(self, task_name):
...
 
 
-class SubmissionStatusHandler(BaseHandler):
+class SubmissionStatusHandler(ContestHandler):
 
     refresh_cookie = False
 
     @tornado.web.authenticated
     @actual_phase_required(0)
-    def get(self, task_name, submission_num):
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/tasksubmission.py]: Codacy Issue found: [Arguments number differs from overridden 'get' method](https://www.codacy.com/app/s-maggiolo/cms/file/2347582228/issues/source?bid=3167167&fileBranchId=3216584#l427)

File: cms/server/contest/handlers/communication.py [hunk 8]
```diff
@@ -38,35 +38,37 @@
...
-class CommunicationHandler(BaseHandler):
+class CommunicationHandler(ContestHandler):
     """Displays the private conversations between the logged in user
     and the contest managers..
 
     """
     @tornado.web.authenticated
-    def get(self):
-        self.set_secure_cookie("unread_count", "0")
+    def get(self, contest_name):
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/communication.py]: Codacy Issue found: [Arguments number differs from overridden 'get' method](https://www.codacy.com/app/s-maggiolo/cms/file/2347582224/issues/source?bid=3167167&fileBranchId=3216584#l53)

File: cms/server/contest/handlers/taskusertest.py [hunk 9]
```diff
@@ -46,30 +46,30 @@
...
-class UserTestInterfaceHandler(BaseHandler):
+class UserTestInterfaceHandler(ContestHandler):
     """Serve the interface to test programs.
 
     """
     @tornado.web.authenticated
     @actual_phase_required(0)
-    def get(self):
+    @multi_contest
+    def get(self, contest_name):
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/taskusertest.py]: Codacy Issue found: [Arguments number differs from overridden 'get' method](https://www.codacy.com/app/s-maggiolo/cms/file/2395198650/issues/source?bid=3167167&fileBranchId=3216584#l68)

File: cms/server/contest/handlers/taskusertest.py [hunk 10]
```diff
@@ -438,16 +441,17 @@ def post(self, task_name):
...
 
-class UserTestStatusHandler(BaseHandler):
+class UserTestStatusHandler(ContestHandler):
 
     refresh_cookie = False
 
     @tornado.web.authenticated
     @actual_phase_required(0)
-    def get(self, task_name, user_test_num):
+    @multi_contest
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/taskusertest.py]: Codacy Issue found: [Arguments number differs from overridden 'get' method](https://www.codacy.com/app/s-maggiolo/cms/file/2395198650/issues/source?bid=3167167&fileBranchId=3216584#l454)

File: cms/server/contest/handlers/main.py [hunk 11]
```diff
@@ -34,38 +34,43 @@
...
 
 
-class MainHandler(BaseHandler):
+class MainHandler(ContestHandler):
     """Home page handler.
 
     """
-    def get(self):
+    @multi_contest
+    def get(self, contest_name):
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/main.py]: Codacy Issue found: [Arguments number differs from overridden 'get' method](https://www.codacy.com/app/s-maggiolo/cms/file/2395198647/issues/source?bid=3167167&fileBranchId=3216584#l59)

File: cms/server/contest/handlers/main.py [hunk 12]
```diff
@@ -256,12 +265,14 @@ def get(self):
...
-    def post(self):
+    @multi_contest
+    def post(self, contest_name):
         participation = self.current_user
 
         if not self.r_params["printing_enabled"]:
-            self.redirect("/")
-            return
+            raise tornado.web.HTTPError(403)
+
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/main.py]: Codacy Issue found: [Line too long (84/80)](https://www.codacy.com/app/s-maggiolo/cms/file/2395198647/issues/source?bid=3167167&fileBranchId=3216584#l275)

File: cms/server/contest/handlers/contest.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/contest.py]: Codacy Issue found: [Line too long (81/80)](https://www.codacy.com/app/s-maggiolo/cms/file/2395198646/issues/source?bid=3167167&fileBranchId=3216584#l288)

File: cms/server/contest/handlers/contest.py [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+        if self.get_secure_cookie(cookie_name) is None:
+            return None
+
+        # Parse cookie.
+        try:
+            cookie = pickle.loads(self.get_secure_cookie(cookie_name))
+            username = cookie[0]
+            password = cookie[1]
+            last_update = make_datetime(cookie[2])
+        except:
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/contest.py]: Codacy Issue found: [No exception type(s) specified](https://www.codacy.com/app/s-maggiolo/cms/file/2395198646/issues/source?bid=3167167&fileBranchId=3216584#l215)

File: cms/server/contest/handlers/communication.py [hunk 15]
```diff
@@ -33,40 +33,47 @@
...
+class CommunicationHandler(ContestHandler):
     """Displays the private conversations between the logged in user
     and the contest managers..
 
     """
     @tornado.web.authenticated
-    def get(self):
-        self.set_secure_cookie("unread_count", "0")
+    @multi_contest
+    def get(self, contest_name):
```
Comment (stefano-maggiolo) [inline cms/server/contest/handlers/communication.py]: Codacy Issue found: [Arguments number differs from overridden 'get' method](https://www.codacy.com/app/s-maggiolo/cms/file/2395198645/issues/source?bid=3167167&fileBranchId=3216584#l56)

**Comment by wil93** [comment]
Seeing that #545 was merged my faith in this PR was restored and I decided to try and rebase it :smile: 

It was quite hard (especially because of these commits: cd09d99558cbc24502411b567e5593228822d825, 8f0ec9f6bfe155c89b328e660168be04c8ade73f and 7442fac94fbb7e350cdc7c562b400d062baaed7a).

Please let's merge this before other commits break CWS again :pray:

File: cms/server/contest/handlers/contest.py [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+        # Run render_params() now, not at the beginning of the request,
+        # because we need contest_name
+        self.r_params = self.render_params()
+
+    def render_params(self):
+        ret = super(ContestHandler, self).render_params()
+
+        ret["contest"] = self.contest
+        ret["contest_root"] = ret["url_root"]
+        ret["real_contest_root"] = "/"
```
Comment (lucach) [inline cms/server/contest/handlers/contest.py]: Can you explain the semantic of `contest_root` and `real_contest_root`?
In some scenarios, this PR causes trouble with `StartHandler` and `LogoutHandler`.

Comment (wil93) [inline cms/server/contest/handlers/contest.py]: `url_root` includes all necessary `../` to go back to the root, so if the URL is `localhost:8888/a/b/c` then `url_root` is `../../`

`real_url_root` is always `/`

`contest_root` is `url_root` + the name of the contest, `real_contest_root` is `/` + the contest name.

Anyway, I once saw some issues with LogoutHandler, but then I think I fixed them. Do you still see issues?

Comment (lucach) [inline cms/server/contest/handlers/contest.py]: Assume that current URL is `localhost:8888/a`. Let's start our timer to compete, actually calling `StartHandler`.
Now that I've understood the meaning, variables are correctly setted as:

- `url_root = ..`
- `contest_root = ../a`
- `real_contest_root = /a`

`StartHandler` tries to redirect the user to `../a` (see https://github.com/wil93/cms/blob/multi_cws/cms/server/contest/handlers/main.py#L148)
but our custom redirector (see https://github.com/wil93/cms/blob/multi_cws/cms/server/util.py#L714) creates the invalid URL `localhost:8888/a/..../a`

Comment (wil93) [inline cms/server/contest/handlers/contest.py]: I thought I had fixed that, but it seems this bug is still on the `multi_cws` branch :confused: 

I think I fixed it in PR #646 (which is based on this one). I'll backport the fixes here :smile:

Comment (wil93) [inline cms/server/contest/handlers/contest.py]: Done, thanks :smile: it looks like I had fixed this in the wrong branch (and just for LogoutHandler, thus leaving a broken StartHandler)

Comment (lucach) [inline cms/server/contest/handlers/contest.py]: Everything works as expected now ;)

**Comment by andreyv** [comment]
With this branch I don't see a language selection option in CWS. Can you reproduce this?

**Comment by wil93** [comment]
@andreyv Yes, I had noticed that some time ago but didn't investigate. I tried to fix it now, it should work

**Comment by stefano-maggiolo** [comment]
I rebased this PR to master - it was great fun, but I suggest we don't do it again and try to close it asap, by commenting here and fixing up in my branch (once done commenting, I can publish my branch to let you try it out, if you want)

Unless you have better ideas, that is...

Looking again at the code now.

**Comment by wil93** [comment]
> try to close it asap, by commenting here and fixing up in my branch

I'm not sure I understand what's the plan 😅

File: cms/server/contest/server.py [hunk 17]
```diff
@@ -83,15 +84,25 @@ def __init__(self, shard, contest):
                               "contest_listen_address and contest_listen_port "
                               "in cms.conf." % __name__)
 
+        self.contest = contest
+        del contest
```
Comment (lerks) [inline cms/server/contest/server.py]: Why are you deleting a local variable?

**Comment by lerks** [comment]
How does this work when CWS is managed by RS? From what I see I understand that CWS will always be started in single-contest mode. Is that correct?

**Comment by wil93** [comment]
If you put -c some_number, then it will start as usual (single contest),
otherwise it will require a prefix /contest_name/ in the URL. In the home
page a list of contests will be shown.

The reason of deleting the local variable was just to ensure a single point
of truth, I think I've seen somebody using that practice

Il mar 28 mar 2017, 15:41 Luca Wehrstedt <notifications@github.com> ha
scritto:

**Comment by lerks** [comment]
You didn't understand my question. I wasn't talking about how to launch CWS from the command line. I was in some sense saying that it appears to me that it's impossible to launch CWS in multi-contest mode when RS is used to manage it. Am I wrong?

Also, AFAIK, we don't do that, we don't remove local variables.

**Comment by stefano-maggiolo** [comment]
@lerks I don't see it as a problem for the first submitted version of this code.

@wil93 I wanted to comment here, but my work is done mostly when I don't have internet connection and reviewable doesn't make it easy - how about I jot down a first fixup commit with straightforward stuff, send a PR with these commits and my fixups, and we can continue the discussion there?

**Comment by wil93** [comment]
Yes I did not modify RS so it will launch CWS as single contest. I don't
use RS (I prefer supervisord) so I didn't think about this.

Stefano: OK for me

Il mar 28 mar 2017, 15:55 Stefano Maggiolo <notifications@github.com> ha
scritto:

**Comment by stefano-maggiolo** [comment]
See #739

**Comment by stefano-maggiolo** [comment]
Damn, just a tad out of our service level agreement of one year ;P

Merged!

**Pull request closed without merging**